autofit/non_linear/parallel/sneaky.py

```diff
@@ -349,3 +350,171 @@ def __del__(self):
+    """
+    FunctionCache.fitness = fitness
+    FunctionCache.prior_transform = prior_transform
+    FunctionCache.fitness_args = fitness_args
+    FunctionCache.fitness_kwargs = fitness_kwargs
+    FunctionCache.prior_transform_args = prior_transform_args
+    FunctionCache.prior_transform_kwargs = prior_transform_kwargs
+
+
+def fitness_cache(x):
+    """
+    Likelihood function call
+    """
+    return FunctionCache.fitness(x, *FunctionCache.fitness_args,
+                                 **FunctionCache.fitness_kwargs)
+
+
+def prior_transform_cache(x):
+    """
+    Prior transform call
+    """
+    return FunctionCache.prior_transform(x, *FunctionCache.prior_transform_args,
+                                         **FunctionCache.prior_transform_kwargs)
+
+class SneakierPool:
+    def __init__(
+            self,
+            processes: int,
+            fitness: Callable,
+            prior_transform: Optional[Callable] = None,
+            fitness_args: Optional[Iterable] = None,
+            fitness_kwargs: Optional[dict] = None,
+            prior_transform_args: Optional[Iterable] = None,
+            prior_transform_kwargs: Optional[dict] = None,
+    ):
+
+        self.fitness_init = fitness
+        self.prior_transform_init = prior_transform
+        self.fitness = fitness_cache
+        self.prior_transform = prior_transform_cache
+        self.fitness_args = fitness_args or ()
+        self.fitness_kwargs = fitness_kwargs or {}
+        self.prior_transform_args = prior_transform_args or ()
+        self.prior_transform_kwargs = prior_transform_kwargs or {}
+        self.processes = processes
+        self.pool = None
+        try:
+            from mpi4py import MPI
+            self.comm = MPI.COMM_WORLD
+        except ModuleNotFoundError:
+            pass
+
+        self._processes = self.comm.size
+        
+        init_args = (
+            self.fitness_init, self.prior_transform_init,
+            self.fitness_args, self.fitness_kwargs,
+            self.prior_transform_args, self.prior_transform_kwargs
+        )
+        initializer(*init_args)
+
+    def check_if_mpi(self):
+
+        return self._processes > 1
+
+    def is_master(self):
+
+        is_mpi = self.check_if_mpi()
+        if is_mpi:
+            return_value = self.comm.rank == 0
+        else:
+            return_value = True
+        
+        return return_value
+
+    def wait(self):
+
+        is_mpi = self.check_if_mpi()
+        if is_mpi:
+            self.pool.wait()
+        else:
+            pass
+            warnings.warn(
+                "Cannot wait for pool to finish if not using MPI"
+            )
+
+    def __enter__(self):
+        """
+        Activate the mp / mpi pool
+        """
+
+        use_mpi = self.check_if_mpi()
+
+        if use_mpi:
+
+            from schwimmbad import MPIPool
+
+            if self.is_master():
+                logger.info("... using Schwimmbad MPIPool")
```

Comment:
What happens if we don't add the master check here? Do we just get really noisy logging?